### ISSUE TITLE
flex attention does not leverage masking, memory error

### ISSUE THREAD

**Issue by Bastien-mva** [pytorch/pytorch#152528]
### 🐛 Describe the bug

I use the `flex_attention`  for very large time series, and want to compute a sliding window attention. I create a `block_mask` to avoid computing the not needed outer products of classical attention. However, it seems that `flex_attention` computes this outer product between query and key. This gives a memory error for large time series. I tried both on GPU and CPU. 

Here is a minimal example: 

```
  import torch

  from torch.nn.attention.flex_attention import flex_attention, create_block_mask

  WINDOW = 5

  batch_size = 16
  nhead = 3
  seq_len = 25000
  head_dim = 32

  queries = torch.randn(batch_size, nhead, seq_len, head_dim).cuda()
  values = torch.randn(batch_size, nhead, seq_len, head_dim).cuda()
  keys = torch.randn(batch_size, nhead, seq_len, head_dim).cuda()

  def sliding_window_mask(b, h, q_idx, kv_idx):
      return q_idx - kv_idx <= WINDOW

  block_mask = create_block_mask(
      sliding_window_mask,
      B=None,
      H=None,
      Q_LEN=seq_len,
      KV_LEN=seq_len,
      device=queries.device,
      _compile=True,
  )

  flex_attention(queries, keys, values, None, block_mask)
```

I thought the whole point of using a mask was not to compute the useless outer products that will not be used after. Am I missing something here ? This should be O(seq_len*window) in memory no ? 

### Versions

Collecting environment information...
PyTorch version: 2.7.0+rocm6.3
Is debug build: False
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: 6.3.42131-fa1d09cbd

OS: Ubuntu 24.04.2 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.12.3 (main, Feb  4 2025, 14:48:35) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.11.0-24-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: AMD Radeon Graphics (gfx1035)
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: 6.3.42131
MIOpen runtime version: 3.3.0
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        48 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               16
On-line CPU(s) list:                  0-15
Vendor ID:                            AuthenticAMD
Model name:                           AMD Ryzen 7 7735HS with Radeon Graphics
CPU family:                           25
Model:                                68
Thread(s) per core:                   2
Core(s) per socket:                   8
Socket(s):                            1
Stepping:                             1
Frequency boost:                      enabled
CPU(s) scaling MHz:                   41%
CPU max MHz:                          4829.0000
CPU min MHz:                          400.0000
BogoMIPS:                             6387.63
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk clzero irperf xsaveerptr rdpru wbnoinvd cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm debug_swap
Virtualization:                       AMD-V
L1d cache:                            256 KiB (8 instances)
L1i cache:                            256 KiB (8 instances)
L2 cache:                             4 MiB (8 instances)
L3 cache:                             16 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-15
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Mitigation; Safe RET
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Retpolines; IBPB conditional; IBRS_FW; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.2.5
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] pytorch-triton-rocm==3.3.0
[pip3] torch==2.7.0+rocm6.3
[pip3] torchaudio==2.7.0+rocm6.3
[pip3] torchvision==0.22.0+rocm6.3
[pip3] triton==3.3.0
[conda] Could not collect

cc @chauhang @penguinwu @zou3519 @ydwu4 @bdhirsh @Chillee @drisspg @yanboliang @BoyuanFeng

**Comment by drisspg** [comment]
So you only see memory and performance of a fused kernel by running with torch.compile

e.g. 
``` Python
  import torch

  from torch.nn.attention.flex_attention import flex_attention, create_block_mask
flex_attention_compiled = torch.compile(flex_attention)

  WINDOW = 5

  batch_size = 16
  nhead = 3
  seq_len = 25000
  head_dim = 32

  queries = torch.randn(batch_size, nhead, seq_len, head_dim).cuda()
  values = torch.randn(batch_size, nhead, seq_len, head_dim).cuda()
  keys = torch.randn(batch_size, nhead, seq_len, head_dim).cuda()

  def sliding_window_mask(b, h, q_idx, kv_idx):
      return q_idx - kv_idx <= WINDOW

  block_mask = create_block_mask(
      sliding_window_mask,
      B=None,
      H=None,
      Q_LEN=seq_len,
      KV_LEN=seq_len,
      device=queries.device,
      _compile=True,
  )

  flex_attention_compiled(queries, keys, values, None, block_mask)
```
